cli/src/main.rs

```diff
@@ -249,29 +249,82 @@ async fn serial(
     let _raw_guard = RawTermiosGuard::stdio_guard()
         .with_context(|| anyhow!("failed to set raw mode"))?;
 
-    let mut stdin = tokio::io::stdin();
     let mut stdout = tokio::io::stdout();
-    let mut next_raw = false;
 
-    loop {
-        tokio::select! {
-            c = stdin.read_u8() => {
-                match c? {
+    // https://docs.rs/tokio/latest/tokio/io/trait.AsyncReadExt.html#method.read_exact
+    // is not cancel safe! Meaning reads from tokio::io::stdin are not cancel
+    // safe. Spawn a separate task to read and put bytes onto this channel.
+    let (tx, mut rx) = tokio::sync::mpsc::channel(16);
+
+    tokio::spawn(async move {
+        let mut stdin = tokio::io::stdin();
+
+        // next_raw must live outside loop, because Ctrl-A should work across
+        // multiple inbuf reads.
+        let mut next_raw = false;
+
+        loop {
+            let mut inbuf = vec![0u8; 1024];
```

Comment:
This could just be a stack allocated array right. We don't really need to create a new vec evry iteration

Reply:
addressed in 34e3329a53193e9bb90867176a732428e1e3fdc6